usr/src/lib/smbclnt/libfknsmb/common/fake_stream.c

```diff
@@ -0,0 +1,1834 @@
+	 * should be no references to bcp since it may be freed by
+	 * runbufcalls(). Since bcp_id field is returned, we save its value in
+	 * the local var.
+	 */
+	bc_id = bcp->bc_id = (bufcall_id_t)(bid += 2);	/* keep it odd */
+
+	/*
+	 * add newly allocated stream event to existing
+	 * linked list of events.
+	 */
+	if (strbcalls.bc_head == NULL) {
+		strbcalls.bc_head = strbcalls.bc_tail = bcp;
+	} else {
+		strbcalls.bc_tail->bc_next = bcp;
+		strbcalls.bc_tail = bcp;
+	}
+
+	cv_signal(&strbcall_cv);
+	mutex_exit(&strbcall_lock);
+	return (bc_id);
+}
+
+/*
+ * Cancel a bufcall request.
+ */
+void
+unbufcall(bufcall_id_t id)
+{
+	strbufcall_t *bcp, *pbcp;
+
+	mutex_enter(&strbcall_lock);
+again:
+	pbcp = NULL;
+	for (bcp = strbcalls.bc_head; bcp; bcp = bcp->bc_next) {
+		if (id == bcp->bc_id)
+			break;
+		pbcp = bcp;
+	}
+	if (bcp) {
+		if (bcp->bc_executor != NULL) {
+			if (bcp->bc_executor != curthread) {
+				cv_wait(&bcall_cv, &strbcall_lock);
+				goto again;
+			}
+		} else {
+			if (pbcp)
+				pbcp->bc_next = bcp->bc_next;
+			else
+				strbcalls.bc_head = bcp->bc_next;
+			if (bcp == strbcalls.bc_tail)
+				strbcalls.bc_tail = pbcp;
+			kmem_free(bcp, sizeof (strbufcall_t));
+		}
+	}
+	mutex_exit(&strbcall_lock);
+}
+
+#endif	/* _KERNEL */
+
+/*
+ * Duplicate a message block by block (uses dupb), returning
+ * a pointer to the duplicate message.
+ * Returns a non-NULL value only if the entire message
+ * was dup'd.
+ */
+mblk_t *
+dupmsg(mblk_t *bp)
+{
+	mblk_t *head, *nbp;
+
+	if (!bp || !(nbp = head = dupb(bp)))
+		return (NULL);
+
+	while (bp->b_cont) {
+		if (!(nbp->b_cont = dupb(bp->b_cont))) {
+			freemsg(head);
+			return (NULL);
+		}
+		nbp = nbp->b_cont;
+		bp = bp->b_cont;
+	}
+	return (head);
+}
+
+#define	DUPB_NOLOAN(bp) \
+	((((bp)->b_datap->db_struioflag & STRUIO_ZC) != 0) ? \
+	copyb((bp)) : dupb((bp)))
+
+mblk_t *
+dupmsg_noloan(mblk_t *bp)
+{
+	mblk_t *head, *nbp;
+
+	if (bp == NULL || DB_TYPE(bp) != M_DATA ||
+	    ((nbp = head = DUPB_NOLOAN(bp)) == NULL))
+		return (NULL);
+
+	while (bp->b_cont) {
+		if ((nbp->b_cont = DUPB_NOLOAN(bp->b_cont)) == NULL) {
+			freemsg(head);
+			return (NULL);
+		}
+		nbp = nbp->b_cont;
+		bp = bp->b_cont;
+	}
+	return (head);
+}
+
+/*
+ * Copy data from message and data block to newly allocated message and
+ * data block. Returns new message block pointer, or NULL if error.
+ * The alignment of rptr (w.r.t. word alignment) will be the same in the copy
+ * as in the original even when db_base is not word aligned. (bug 1052877)
+ */
+mblk_t *
+copyb(mblk_t *bp)
+{
+	mblk_t	*nbp;
+	dblk_t	*dp, *ndp;
+	uchar_t *base;
+	size_t	size;
+	size_t	unaligned;
+
+	ASSERT(bp->b_wptr >= bp->b_rptr);
+
+	dp = bp->b_datap;
+	if (dp->db_fthdr != NULL)
+		STR_FTEVENT_MBLK(bp, caller(), FTEV_COPYB, 0);
+
+	/*
+	 * Special handling for Multidata message; this should be
+	 * removed once a copy-callback routine is made available.
+	 */
+	if (dp->db_type == M_MULTIDATA) {
+#ifdef	_KERNEL
```

Comment:
Kill it with fire.  In fact I'd just nuke this whole block... M_MULTIDATA will never ever be seen in your user space code.  It only happens for certain legacy SPARC drivers. (Cassini.)